### PR TITLE
[int64] Support for HP / Mana / End / Damage / Hate

### PR DESCRIPTION
**Pull request by Akkadius**
Need to do additional testing, also looking for others to do testing as well

**Implemented**

* HP
* Mana
* Endurance
* Damage
* Hate

**What's Verified Working**

* HP
* Mana
* Damage
* Hate
* Spell Damage

Damage works but clients only display int32 so we'll have to come up with some sort of way to emulate messages that overflow. The client just displays it as a "miss" even though the server still registers it. That will be outside of the scope of this PR though

image

### CODE REVIEW

File: zone/hate_list.cpp [hunk 1]
```diff
@@ -745,7 +745,7 @@ void HateList::SpellCast(Mob *caster, uint32 spell_id, float range, Mob* ae_cent
 	//crashes when people kick the bucket in the middle of this call
 	//that invalidates our iterator but there's no way to know sadly
 	//So keep a list of entity ids and look up after
-	std::list<uint32> id_list;
+	std::list<uint64> id_list;
 	range = range * range;
```
Comment (Kinglykrab) [inline zone/hate_list.cpp:749]: Intentional change here?

File: zone/perl_mob.cpp [hunk 2]
```diff
@@ -3636,12 +3636,12 @@ XS(XS_Mob_SetHate); /* prototype to pass -Wmissing-prototypes */
...
 	dXSARGS;
 	if (items < 2 || items > 4)
-		Perl_croak(aTHX_ "Usage: Mob::SetHate(THIS, Mob* other, [int32 hate = 0], [int32 damage = 0])"); // @categories Hate and Aggro
+		Perl_croak(aTHX_ "Usage: Mob::SetHate(THIS, Mob* other, [int32 hate = 0], [int64 damage = 0])"); // @categories Hate and Aggro
 	{
 		Mob *THIS;
 		Mob *other;
 		int32 hate;
-		int32 damage;
+		int64 damage;
```
Comment (Kinglykrab) [inline zone/perl_mob.cpp:3644]: Hate, too? Still casting to int32 for both below as well.

Comment (Akkadius) [reply zone/perl_mob.cpp:3644]: Definitely not done yet hence the draft, still plan on making sure all the edges are touched

File: zone/hate_list.cpp (hunk 1 above)
Comment (Akkadius) [reply zone/hate_list.cpp:749]: Nope, won't hurt anything but take 4 more bytes of register memory per entry. I'll probably bring this one back down as I work through this more in a bit.

Comment (Kinglykrab) [reply zone/hate_list.cpp:749]: No worries, just making note since it's spell ID and stuff that had changed.

File: zone/perl_mob.cpp (hunk 2 above)
Comment (Kinglykrab) [reply zone/perl_mob.cpp:3644]: No worries. As far as I'm aware, int64 does not function with the Perl API. See $client->GetCarriedMoney() with over 2,000,000 Platinum for example.

Comment (Akkadius) [reply zone/perl_mob.cpp:3644]: Perl should have support given 64 bit installation / libs

https://stackoverflow.com/questions/31490111/using-64-bits-integers-with-perl-xs

That's not to say we might need to do additional investigation

**Comment by Kinglykrab** [comment]
**Testing Notes**
- Health doesn't seem to 100% line up in display messages.
image
image
- `hp` is set to `275,000,000,000`
- `mana` is set to `275,000,000,000`
- `hp_regen_rate` is set to `10,000,000,000`
- `hp_regen_per_second` is set to `10,000,000,000`
- `mana_regen_rate` is set to `10,000,000,000`
- #damage seems to be capping damage in message as well, used #damage `10,000,000,000` here.
- Using #damage `100,000,000,000` the NPC does go down `100,000,000,000`.
- Guess we need to send our own messages with the same filter if damage > 2,000,000,000.
image
- Spells seem to cap at `2,147,483,648` damage even with these changes and a `spellscale` of `2,000,000,000`.
image
image
- Minimum and maximum NPC damage are still `int`.

**Perl API**
- `$hate_entry->GetHate()` still using `int` for return.
- `$hate_entry->GetDamage()` still using `int` for return.
- `$mob->AddToHateList()` still using `int` for hate and damage in casting.
- `$mob->CheckAggroAmount()` still using `uint32` for return.
- `$mob->CheckHealAggroAmount()` still using `uint32` for return.
- `$mob->DoArcheryAttackDmg()` still using `uint16` for damage.
- `$mob->DoMeleeSkillAttackDmg()` still using `uint16` for damage.
- `$mob->DoSpecialAttackDamage()` still using `int` for minimum/maximum damage and hate override.
- `$mob->DoThrowingAttackDmg()` still using `uint16` for damage.
- `$mob->GetHateAmount()` still using `uint32` for return.
- `$mob->GetDamageAmount()` still using `uint32` for return.
- `$mob->SetHate()` still using `int` for hate and damage in casting.

**Lua API**
- `mob:DoArcheryAttackDmg()` still using` int` for damage.
- `mob:DoMeleeSkillAttackDmg()` still using` int` for damage.
- `mob:DoSpecialAttackDamage()` still using` int` for minimum/maximum damage and hate override.
- `mob:DoThrowingAttackDamage()` still using` int` for damage.
- `mob:GetDamageAmount()` is still using `uint32` in the Luabind definition.

The Archery, Melee, Special, and Throwing may be able to be ignored.

**Comment by Kinglykrab** [comment]
With Perl AddToHateList with int64 for hate/damage I was able to get hate and damage to be working fine over int cap.
image

**Comment by Akkadius** [comment]
Have a pet casting damage above int32 but it suffers from the same problem as physical damage, it shows as a miss to the client

image

image

**Comment by Akkadius** [comment]
Pets actually fully showing up

image

File: zone/zonedump.h [hunk 3]
```diff
@@ -104,7 +104,7 @@ struct NPCType
 	uint8	prim_melee_type;
 	uint8	sec_melee_type;
 	uint8	ranged_type;
-	int32	hp_regen;
+	int64	hp_regen;
 	int64	hp_regen_per_second;
 	int32	mana_regen;
```
Comment (Kinglykrab) [inline zone/zonedump.h:109]: Should this be int64?

File: zone/perl_mob.cpp [hunk 4]
```diff
@@ -3579,12 +3579,12 @@ XS(XS_Mob_AddToHateList); /* prototype to pass -Wmissing-prototypes */
 XS(XS_Mob_AddToHateList) {
 	dXSARGS;
 	if (items < 2 || items > 7)
-		Perl_croak(aTHX_ "Usage: Mob::AddToHateList(THIS, Mob* other, [int32 hate = 0], [int32 damage = 0], [bool yell_for_help = true], [bool frenzy = false], [bool buff_tic = false])"); // @categories Hate and Aggro
+		Perl_croak(aTHX_ "Usage: Mob::AddToHateList(THIS, Mob* other, [int32 hate = 0], [int64 damage = 0], [bool yell_for_help = true], [bool frenzy = false], [bool buff_tic = false])"); // @categories Hate and Aggro
```
Comment (Kinglykrab) [inline zone/perl_mob.cpp:3582]: int32 in message should be int64.

File: zone/mob.h [hunk 5]
```diff
@@ -1422,12 +1437,12 @@ class Mob : public Entity {
...
-	int32 hp_regen;
+	int64 current_hp;
+	int64 max_hp;
+	int64 base_hp;
+	int64 current_mana;
+	int64 max_mana;
+	int64 hp_regen;
 	int64 hp_regen_per_second;
 	int32 mana_regen;
 	int32 ooc_regen;
```
Comment (Kinglykrab) [inline zone/mob.h:1448]: int64 here?

Comment (Akkadius) [reply zone/mob.h:1448]: `hp_regen` isn't fully using int64 so I didn't touch OOC fully

Only `hp_regen_per_second` is int64

File: zone/zonedump.h (hunk 3 above)
Comment (Akkadius) [reply zone/zonedump.h:109]: It's fine, only fields I'm fully supporting int64 out of the DB currently

```sql
ALTER TABLE npc_types MODIFY COLUMN hp BIGINT;
ALTER TABLE npc_types MODIFY COLUMN mana BIGINT;
ALTER TABLE npc_types MODIFY COLUMN hp_regen_rate BIGINT;
ALTER TABLE npc_types MODIFY COLUMN mana_regen_rate BIGINT;
```

File: zone/perl_mob.cpp (hunk 4 above)
Comment (Akkadius) [reply zone/perl_mob.cpp:3582]: Good 👀

File: zone/zonedump.h (hunk 3 above)
Comment (Kinglykrab) [reply zone/zonedump.h:109]: So mana_regen isn't mana_regen_rate?

File: zone/mob.h (hunk 5 above)
Comment (Kinglykrab) [reply zone/mob.h:1448]: Okay, should mana_regen be int64 there?

Comment (Akkadius) [reply zone/mob.h:1448]: I thought you were pointing at ooc_regen, I can scoop these up real quick

Comment (Kinglykrab) [reply zone/mob.h]: Didn't know if both or just one/other was supposed to be. Thanks. 👍

**Review by Kinglykrab — APPROVED** [review]

**Comment by Kinglykrab** [comment]
👍

**Pull request merged**